Add documentation for starting and exiting (#12645)
nvaccess/nvda: devDocs/startupShutdown.md

    1. Via the NVDA menu -> Exit
        - test: `startupShutdownNVDA.Quits from menu`
1. A process sending `WM_QUIT`, eg a new NVDA process starting
1. A handled crash (directly causes a new process to start, terminates unsafely)
    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
1. An unhandled crash (terminates unsafely)
    - requires manual testing/confirmation
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)

## Manual testing
Instructions for testing startup / shutdown.

### Start from shortcut
Prerequisites:
 - NVDA installed
 - Shortcut enabled during installation

Steps:
 1. Press (or emulate) Ctrl+Alt+N, observe NVDA starts up

Variation:
- At step 1. A version of NVDA is already running. Observe running version exists before installed version starts up.

### Windows Sign-in screen, automatic start
Prerequisites:
 - NVDA installed
 - Enable "Use NVDA during sign-in"

Steps:
 1. Start Windows
 1. Observe NVDA announces the Windows sign-in screen

### UAC, automatic start
Prerequisites:
 - NVDA installed

> Steps:

Author: I'm not sure how to trigger this, or the full prerequisites.

Reviewer: No, I'm not entirely sure either. All I know is that when NVDA is started it follows the process defined [here](https://docs.microsoft.com/en-us/windows/win32/winauto/ease-of-access---assistive-technology-registration#windows-logo-key--u), it sets a registry key in ease of access, and sends the key `windows + u`. Windows then attempts to start NVDA with the `--ease-of-access` param.

@michaelDCurran might be able to provide more information about the expectations here.

    1. An input gesture to restart
    1. After changing some settings (eg installed add-ons or UI language), user prompted on dialog exit.
    1. Via the NVDA menu -> Exit
        - test: `startupShutdownNVDA.Quits from menu`
1. A process sending `WM_QUIT`, eg a new NVDA process starting
1. A handled crash (directly causes a new process to start, terminates unsafely)
    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
1. An unhandled crash (terminates unsafely)
    - requires manual testing/confirmation
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)

## Manual testing
Instructions for testing startup / shutdown.

### Start from shortcut
Prerequisites:
 - NVDA installed
 - Shortcut enabled during installation

Steps:
 1. Press (or emulate) Ctrl+Alt+N, observe NVDA starts up

Variation:
- At step 1. A version of NVDA is already running. Observe running version exists before installed version starts up.

### Windows Sign-in screen, automatic start
Prerequisites:
 - NVDA installed
 - Enable "Use NVDA during sign-in"

Steps:

>  1. Start Windows

Reviewer: Is there an easier way, assume the tester is already running windows.
They can probably "sign out", right?

Author: Yes, assuming signing out always exits the running instance and starts a new one on the sign in screen. I specified this as I wasn't certain of that behaviour, and it might be confused with locking Windows, which isn't a valid pathway.

Author: ```suggestion
 1. Sign out (not lock) Windows
```

Reviewer: Yes, when you sign out I believe you are transferred to a different "desktop" a "secure desktop" owned by the "system". NVDA has to start again.

>  1. Start Windows

Author: ```suggestion
 1. Sign out (not lock) Windows
```

 1. Observe NVDA announces the Windows sign-in screen

### UAC, automatic start
Prerequisites:
 - NVDA installed

Steps:

### Windows Successful sign-in, automatic start
Prerequisites:
 - NVDA installed
 - Enable "Start NVDA after I sign in"

Steps:
 1. Start Windows
 1. Sign in
 1. Observe NVDA starts

### Running the *.exe

Steps:

>  1. Run the NVDA exe from cmd

Reviewer: Can you make this more explicit, eg 
```
1. press `win+r`
1. enter <path to nvda.exe>
1. press enter
```

Author: ```suggestion
 1. Press `win+r`
 1. Enter <path to nvda.exe>
 1. Press enter
```

 1. Observe NVDA starts

Variation:

> - installed copy, portable copy, or installer

Reviewer: How are these variations achieved? Can an alpha tester follow this?

Author: ```suggestion
- using an installer .exe, an installed copy .exe, or a portable copy .exe
   - the installer allows you to create an installed copy and a portable copy
```

> - CLI options

Reviewer: What options?

Author: Well any valid combination of the [ones we accept](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#CommandLineOptions) should be tested, this would be quite a lot of work though. 

Are you suggesting I should link the options or describe testing pathways for each combination?

Reviewer: > describe testing pathways for each combination?

No, I don't think that would be relevant here. This should ideally be a guide specific to manually testing the startup / shutdown logic.

However, if there are specific options that are relevant, those should be included. If not, just remove this item.

Author: ```suggestion
```

### Running from source (runnvda.bat)

> 

Author: ```suggestion
Prerequisites
- clone project and build NVDA (see [project readme](https://github.com/nvaccess/nvda/blob/master/readme.md#getting-the-source-code)).

```

Steps:

>  1. Run `runnvda.bat` from cmd

Reviewer: Prerequisites: clone project and build NVDA (see project readme).

- From internally calling [triggerNVDAExit](#When-exiting-from-triggerNVDAExit)
- Receiving [WM_QUIT](#When-exiting-from-WM_QUIT) Windows message
- Receiving [wx.EVT_END_SESSION](#When-exiting-from-wxEVT_END_SESSION) due to Windows session ending

### When exiting from `triggerNVDAExit`
* Called from within NVDA.
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated
    1. addons and the brailleViewer are terminated, so we can close all windows safely
    1. All wx windows are closed
    1. Now that windows are closed, a new NVDA instance is started if requested

### When exiting from `WM_QUIT`
* [A Windows Message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) received from an external process, such as another NVDA process.
* NVDA accepts `WM_QUIT` messages from other processes and creates a [named window](https://docs.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window#creating-the-window) that can be discovered.

> * `WM_QUIT` is handled by `wx`, which force closes all wx windows (other UI features like the systray icon are not windows, and remain) and then exits the main loop. `triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT````

Reviewer: There are too many grave characters on the end of WM_QUIT

Author: ```suggestion
* `WM_QUIT` is handled by `wx`, which force closes all wx windows (other UI features like the systray icon are not windows, and remain) and then exits the main loop.
`triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT`
```

* `WM_QUIT` is handled by `wx`, which force closes all wx windows (other UI features like the systray icon are not windows, and remain) and then exits the main loop.
`triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT`
* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
* Using a custom message has been considered:
  - Would allow custom handling (eg just `triggerNVDAExit`)
  - Unfortunately, older NVDA versions will only be aware of `WM_QUIT`, so we'd need to send `WM_QUIT` to these versions.
  - Sending the custom message, waiting for a timeout, then sending `WM_QUIT` adds a significant wait time
  - Identifying the running version (to selectively send the message) requires maintaining 2 message windows in NVDA (one for legacy behaviour) and adds complexity

### When exiting from `wx.EVT_END_SESSION`
* This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
* On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
* Other actions are not performed as we have limited time to perform an action for this event.

> NVDA is expected to run as long as possible during the sign out process.

Author: ```suggestion
NVDA is expected to run as long as possible during the sign out process.
This is achieved through the Windows API, using [SHUTDOWN_NORETRY](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters)
```

Reviewer: I think the important part (for shutdown order) is the `dwLevel` parameter: 

> The shutdown priority for a process relative to other processes in the system. The system shuts down processes from high dwLevel values to low. 

> dwLevel of `100-1FF` application reserved last shutdown range.

Reviewer: I think it would be worth pointing that out.

Reviewer: Perhaps even note that value 0x100 is used as the lowest value reserved for non system applications

# NVDA Starting and Exiting outline

## Ways to start NVDA:

1. For an installed copy:
    1. Ctrl+Alt+N (Desktop shortcut)
        - test: `startupShutdownNVDA.Starts from desktop shortcut`
    1. Automatically via Ease of Access on the Windows sign-in screen (at boot or signing out of a previous session)
    1. Automatically via Ease of Access on User Account Control (UAC) screens
    1. Automatically by Ease of Access after signing in to Windows
1. For an installed copy, portable copy, installer:
    1. An exiting instance of NVDA starting a new process (see shutting down procedures)

>     1. By running the exe (see cli options too).

Reviewer 2: What "cli options"? This probably should link to the apropriate section of the user guide.

Reviewer: I think we determined elsewhere that "cli options" could be removed.

Author: ```suggestion
    1. By running the exe
```

 - NVDA installed
 - Enable "Start NVDA after I sign in"

Steps:
 1. Start Windows
 1. Sign in
 1. Observe NVDA starts

### Running the *.exe

Steps:
 1. Press `win+r`
 1. Enter <path to nvda.exe>
 1. Press enter
 1. Observe NVDA starts

Variation:
- using an installer (launcher)
   -  eg: `C:\Users\username\Downloads\nvda_2021.1.exe`
- using an installed copy
   - just type `nvda` in place of the .exe
- using a portable copy
   - find and use the path to `nvda.exe`, located within the portable copy directory
   - the installer allows you to create an installed copy and a portable copy

### Running from source (runnvda.bat)
Prerequisites
- clone project and build NVDA (see [project readme](https://github.com/nvaccess/nvda/blob/master/readme.md#getting-the-source-code)).

Steps:
 1. Run `runnvda.bat` from cmd
 1. Observe NVDA starts

### An input gesture to restart

Prerequisite:
- Input gesture to restart is assigned to a key

Steps:
 1. Press (or emulate) the input gesture
 1. Observe that NVDA exits
 1. Observe that a new instance is started

## Technical notes

These notes are aimed at developers, wishing to understand technical aspects of the NVDA start and exit.

> 1. No more than one NVDA process instance should be running time. Interactions with itself could cause severe issues, some (non-exhaustive list) examples of sub-systems where this would be a problem:

Reviewer 2: This seems incomplete and should probably say:
```suggestion
1. No more than one NVDA process instance should be running at the same time. Interactions with itself could cause severe issues, some (non-exhaustive list) examples of sub-systems where this would be a problem:
```

\or something similar

   - NVDA config files
   - Global (OS level) keyboard hook
   - Changed / incompatible in-process code
2. As such, we want to be able to detect running instances, cause them to exit, and confirm they have exited.

### Exit hooks/triggers

There are 3 ways that NVDA receives a request to exit:

- From internally calling [triggerNVDAExit](#When-exiting-from-triggerNVDAExit)
- Receiving [WM_QUIT](#When-exiting-from-WM_QUIT) Windows message
- Receiving [wx.EVT_END_SESSION](#When-exiting-from-wxEVT_END_SESSION) due to Windows session ending

### When exiting from `triggerNVDAExit`
* Called from within NVDA.
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated

>     1. addons and the brailleViewer are terminated, so we can close all windows safely

Reviewer 2: Looking at the current code in `core.py` this seems not to be entirely true since only globalPlugins are  terminated. When reading this my expectation would be that all add-ons are terminated and code which does not belong into an add-on i.e. global plugins in the scratchpad is not both of these are not true since as established above only global plugins are terminated and we do not differentiate between one in scratchpad and one bundled in add-ons. Another question which someone reading this may ask is "What does it mean to terminate an add-on?" There is no class from which all add-ons should inherit and `addonHandler`.terminate` is a no-op.

Author: ```suggestion
    1. globalPlugins and the brailleViewer are terminated, so we can close all windows safely
```

`triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT`
* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
* Using a custom message has been considered:
  - Would allow custom handling (eg just `triggerNVDAExit`)
  - Unfortunately, older NVDA versions will only be aware of `WM_QUIT`, so we'd need to send `WM_QUIT` to these versions.
  - Sending the custom message, waiting for a timeout, then sending `WM_QUIT` adds a significant wait time
  - Identifying the running version (to selectively send the message) requires maintaining 2 message windows in NVDA (one for legacy behaviour) and adds complexity

### When exiting from `wx.EVT_END_SESSION`
* This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
* On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
* Other actions are not performed as we have limited time to perform an action for this event.
    * NVDA is expected to run as long as possible during the sign out process.

>     * This is achieved through the Windows API, using [SHUTDOWN_NORETRY](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters).

Author: ```suggestion
    * This is achieved through the [Windows API](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters), by setting the shutdown priority to the lowest reserved value for non-system applications, `0x100`.
```

> NVDA is expected to run as long as possible during the sign out process.
> This is achieved through the Windows API, using [SHUTDOWN_NORETRY](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters)
> With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.

Author: ```suggestion
    * NVDA is expected to run as long as possible during the sign out process.
    * This is achieved through the Windows API, using [SHUTDOWN_NORETRY](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters).
    * With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.
```

### Replacing an existing NVDA instance

With the requirement to only allow a single instance of NVDA, a new NVDA process must be able to replace an existing NVDA process.

> NVDA will exit correctly in response to a `[`WM_QUIT`](#When-exiting-from-WM_QUIT) Windows message, but the process must first be detected / identified in order to send the message.

Author: ```suggestion
NVDA will exit correctly in response to a [`WM_QUIT`](#When-exiting-from-WM_QUIT) Windows message, but the process must first be detected / identified in order to send the message.
```

1. For an installed copy:
    1. Ctrl+Alt+N (Desktop shortcut)
        - can be tested by:
            1. Set the NVDA config to ensure the welcome dialog loads on startup
            1. Have an existing NVDA process running with a known handle
            1. Press (or emulate) Ctrl+Alt+N
            1. Ensure the known process ends

>             1. Wait until the welcome dialog loads

Reviewer: Please add the detailed test plan to a new section. It's useful to to have an over-view for quickly mapping out the user processes, and a details for exactly how to test. This test procedure seems very programmatic, instead imagine writing instructions for a non-programmer human. Likely very much like the steps-to-reproduce you'll fins on an issue.

I'd put it before `## Technical notes`:

```
## Manual testing
Instructions for testing startup / shutdown.

### Start from shortcut
Prerequisites:
 - NVDA installed
 - Shortcut enabled during installation

Steps:
 1. Press (or emulate) Ctrl+Alt+N, observe NVDA starts up

Variation:
- At step 1. A version of NVDA is already running. Observe running version exists before installed version starts up.

### Windows Sign-in screen, automatic up
  

## Technical notes
```

1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)

## Technical notes

These notes are aimed at developers, wishing to understand technical aspects of the NVDA start and exit.

> 1. We want only one NVDA instance running at a time, so that it can't interact with itself.

Author: Would you be able to write on this? I'm not really aware of these mechanisms/problems.

> 1. We want only one NVDA instance running at a time, so that it can't interact with itself.

Reviewer: You can pretty much just add those three points:
```suggestion
1. No more than one NVDA process instance should be running time. Interactions with itself could cause severe issues, some (non-exhaustive list) examples of sub-systems where this would be a problem:
   - NVDA config files
   - Global (OS level) keyboard hook
   - Changed / incompatible in-process code

```

    1. All wx windows are closed
    1. Now that windows are closed, a new NVDA instance is started if requested

### When exiting from `WM_QUIT`
* [A Windows Message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) received from an external process, such another NVDA process.
* NVDA accepts [WM_QUIT](#When-exiting-from-WM_QUIT) messages from other processes and creates a [named window](https://docs.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window#creating-the-window) that can be discovered.

> * Will force the main loop to exit and close most wx Windows.

Author: The system tray icon is an example

> * Will force the main loop to exit and close most wx Windows.

Reviewer: Please add this as an example exception.

> * Will force the main loop to exit and close most wx Windows.

Author: ```suggestion
* Will force the main loop to exit and close most wx windows, but not other GUI objects like the system tray icon.
```

> * We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.

Author: `WM_QUIT` is handled by `wx`, which force closes all wxWindows (which other UI features like the systrayIcon is not) and then exits the mainloop. `triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT`

> * We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.

Reviewer: Please add this explanation to the docs.

> * Changing to using a custom message, which would allow a custom handling (eg just `triggerNVDAExit`), requires compatibility with messaging older NVDA versions that are only aware of `WM_QUIT`.

Reviewer: ```suggestion
* Using a custom message has been considered:
  - Would allow custom handling (eg just `triggerNVDAExit`)
  - Unfortunately, older NVDA versions will only be aware of `WM_QUIT`, so we'd need to send `WM_QUIT` to these versions.
  - Sending the custom message, waiting for a timeout, then sending `WM_QUIT` was tried but....?
  - Identifying the running version (to selectively send the message) was unsuccessful because ...?

```

This point about using a custom message instead of WM_QUIT could probably be under its own heading. There is plenty to discuss about the things tried and the obstacles.

Author: I think work here was mostly limited due to it being unnecessary and adding scope. I've added what notes I could:

```suggestion
* Using a custom message has been considered:
  - Would allow custom handling (eg just `triggerNVDAExit`)
  - Unfortunately, older NVDA versions will only be aware of `WM_QUIT`, so we'd need to send `WM_QUIT` to these versions.
  - Sending the custom message, waiting for a timeout, then sending `WM_QUIT` adds a significant wait time
  - Identifying the running version (to selectively send the message) requires maintaining 2 message windows in NVDA (one for legacy behaviour) and adds complexity
```

### When exiting from `wx.EVT_END_SESSION`
* This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
* On `wx.EVT_END_SESSION`, we save the config and play the exit sound.

> * Other actions are not performed as we have limited time to perform an action for this event. With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.

Reviewer: Could you please clarify the final sentence here?
- Is NVDA expected to run in the blocked shutdown dialog?
- How does NVDA ensure that it is still running, or ensure that it is killed off last?

Also how much time does NVDA have?

Author: >Is NVDA expected to run in the blocked shutdown dialog?

It tends to, until it is killed off last

>Also how much time does NVDA have?

I think this varies on the system

>How does NVDA ensure that it is still running, or ensure that it is killed off last?

I'm not sure here - @michaelDCurran?

Author: ```suggestion
* Other actions are not performed as we have limited time to perform an action for this event.
NVDA is expected to run as long as possible during the sign out process.
With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.
```

Reviewer: Rather than "we", prefer "NVDA" or "the process"

> It tends to, until it is killed off last

What makes it last? Is there code in NVDA that controls this?

### Replacing an existing NVDA instance

> To ensure only one NVDA instance is running, we need to be able to find and end a running process.
> So that a new NVDA process can find a running NVDA process, a named [message window](https://docs.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window#creating-the-window) is created.
> A running NVDA process can be found by using the Windows API to search for the named window.
> A new NVDA process searches for an existing NVDA window, and if it is detected, sends `WM_QUIT`.

Reviewer: ```suggestion
With the requirement to only allow a single instance of NVDA, a new NVDA process must be able to replace an existing NVDA process.
NVDA will exit correctly in response to a `[`WM_QUIT`](#When-exiting-from-WM_QUIT) Windows message, but the process must first be detected / identified in order to send the message.
For new NVDA process to detect an existing NVDA process, a named [message window](https://docs.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window#creating-the-window) is used.
A new NVDA process searches for an existing NVDA window, and if it is detected, sends `WM_QUIT`.
```

The message window is created late during the start up, and destroyed early in exit and is not perfectly indicative of whether or not an NVDA process is running.
As such, we have a [MutEx](#MutEx) that ensures a newly started process blocks until any previous NVDA has finished exiting.

### MutEx

To confirm that another NVDA process is not running,

> a [MutEx](https://docs.microsoft.com/en-us/windows/win32/sync/mutex-objects) is owned by the NVDA process.

Reviewer: At the risk of spelling out the obvious, perhaps add to this something to say:
- NVDA will be blocked from starting until it can acquire the mutex.
- If it can not acquire the mutex within a timeout, startup is aborted.

> Note detects unusual shutdown of prior instance via 

Reviewer: This isn't quite finished

### Running the *.exe

Steps:
 1. Press `win+r`
 1. Enter <path to nvda.exe>
 1. Press enter
 1. Observe NVDA starts

Variation:

> - using an installer .exe, an installed copy .exe, or a portable copy .exe

Reviewer: With an installed copy you can just type `nvda`. 

For the installer and portable copy an example would be helpful. A common question might be which exe? In a portable copy there are:
- `nvda.exe`
- `nvda_eoaProxy.exe`
- `nvda_noUIAccess.exe`
- `nvda_slave.exe`
- `nvda_uiAccess.exe`

For the installer perhaps add "(launcher)" since it also gets called that, and a common example path:
- `C:\Users\username\Downloads\nvda_2021.1.exe`

Author: ```suggestion
- using an installer (launcher)
   -  eg: `C:\Users\username\Downloads\nvda_2021.1.exe`
- using an installed copy
   - just type `nvda` in place of the .exe
- using a portable copy
   - find and use the path to `nvda.exe`, located within the portable copy directory
```
